prov/efa/src/rxr/rxr_pkt_type_req.c

```diff
@@ -2498,37 +2611,75 @@ int rxr_pkt_proc_compare_rta(struct rxr_ep *ep, struct rxr_pkt_entry *pkt_entry)
 	 */
 	if (dt == FI_INT128) {
 		for (i = 0; i < rx_entry->iov_count; ++i) {
-			ofi_int128_t src, cmp;
-			memcpy(&src, src_data + offset, sizeof(ofi_int128_t));
-			memcpy(&cmp, cmp_data + offset, sizeof(ofi_int128_t));
-
-			ofi_atomic_swap_handler(op, dt, rx_entry->iov[i].iov_base,
-									&src,
-									&cmp,
-									rx_entry->atomrsp_data + offset,
-									rx_entry->iov[i].iov_len / dtsize);
-			offset += rx_entry->iov[i].iov_len;
+			efa_mr = (struct efa_mr*) ofi_mr_map_get(&ep->util_ep.domain->mr_map, (rxr_get_rta_hdr(pkt_entry->pkt)->rma_iov + i)->key);
+			hmem_iface = efa_mr->peer.iface;
+
+			if (hmem_iface == FI_HMEM_SYSTEM) {
+				ofi_int128_t src, cmp;
+				memcpy(&src, src_data + offset, sizeof(ofi_int128_t));
+				memcpy(&cmp, cmp_data + offset, sizeof(ofi_int128_t));
+
+				ofi_atomic_swap_handler(op, dt, rx_entry->iov[i].iov_base,
+				                        &src,
+				                        &cmp,
+				                        rx_entry->atomrsp_data + offset,
+				                        rx_entry->iov[i].iov_len / dtsize);
+				offset += rx_entry->iov[i].iov_len;
+			} else {
+				err = rxr_compare_atomic_hmem(efa_mr, &rx_entry->iov[i], src_data + offset,
```

Comment:
Just realized, this will not work for 128, because `src_data` and `cmp_data` is not aligned on 128 bits boundary. You should have hit a segfault with release build.

for 128, you will need to copy `src_data`  and `cmp_data` to a temporary buffer too.